### PR TITLE
feat(image): customer podman host or socket option

### LINKED ISSUE #3098: Support for rootless podman
Rootful podman has it's socket at:
`/run/podman/podman.sock`

But rootless podman has it's socket at:
`/run/user/1000/podman/podman.sock`

So it errors out.

Edit:
Probably using
`export CONTAINERD_ADDRESS=/run/user/1000/podman/podman.sock`
works.

### PR DESCRIPTION
**Pull request by parvez0**
## Description
This pull request adds a new feature option `--podman-host` to the image command. This option allows users to specify a custom Podman host.

command output before introducing this feature
```sh
$ trivy image --image-src podman nginx:latest
2024-03-02T16:49:02.943Z	INFO	Need to update DB
2024-03-02T16:49:02.944Z	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db
2024-03-02T16:49:02.944Z	INFO	Downloading DB...
43.58 MiB / 43.58 MiB [--------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 7.43 MiB p/s 6.1s
2024-03-02T16:49:10.272Z	INFO	Vulnerability scanning is enabled
2024-03-02T16:49:10.272Z	INFO	Secret scanning is enabled
2024-03-02T16:49:10.272Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-03-02T16:49:10.273Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.49/docs/scanner/secret/#recommendation for faster secret detection
2024-03-02T16:49:10.279Z	FATAL	image scan error: scan error: unable to initialize a scanner: unable to initialize an image scanner: 1 error occurred:
	* podman error: unable to initialize Podman client: no podman socket found: stat /run/user/1000/snap.trivy/podman/podman.sock: no such file or directory
```
command output after introducing the feature
```sh
$ trivy image --image-src podman --podman-host /run/user/1000/podman/podman.sock docker.io/nginx:latest
2024-03-02T17:03:12.479Z	INFO	Vulnerability scanning is enabled
2024-03-02T17:03:12.479Z	INFO	Secret scanning is enabled
2024-03-02T17:03:12.479Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-03-02T17:03:12.480Z	INFO	Please see also https://aquasecurity.github.io/trivy/dev/docs/scanner/secret/#recommendation for faster secret detection
2024-03-02T17:03:25.884Z	INFO	Java DB Repository: ghcr.io/aquasecurity/trivy-java-db:1
2024-03-02T17:03:25.884Z	INFO	Downloading the Java DB...
509.15 MiB / 509.15 MiB [------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 18.21 MiB p/s 28s
2024-03-02T17:03:55.181Z	INFO	The Java DB is cached for 3 days. If you want to update the database more frequently, the '--reset' flag clears the DB cache.
2024-03-02T17:03:55.631Z	INFO	Detected OS: debian
2024-03-02T17:03:55.631Z	INFO	Detecting Debian vulnerabilities...
2024-03-02T17:03:55.730Z	INFO	Number of language-specific files: 0

docker.io/nginx:latest (debian 12.5)

Total: 146 (UNKNOWN: 1, LOW: 84, MEDIUM: 32, HIGH: 27, CRITICAL: 2)

┌────────────────────┬─────────────────────┬──────────┬──────────────┬─────────────────────────┬───────────────┬──────────────────────────────────────────────────────────────┐
│      Library       │    Vulnerability    │ Severity │    Status    │    Installed Version    │ Fixed Version │                            Title                             │
├────────────────────┼─────────────────────┼──────────┼──────────────┼─────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────┤
│ apt                │ CVE-2011-3374       │ LOW      │ affected     │ 2.6.1                   │               │ It was found that apt-key in apt, all versions, do not       │
│                    │                     │          │              │                         │               │ correctly...                                                 │
│                    │                     │          │              │                         │               │ https://avd.aquasec.com/nvd/cve-2011-3374                    │
├────────────────────┼─────────────────────┤          │              ├─────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────┤
│ bash               │ TEMP-0841856-B18BAF │          │              │ 5.2.15-2+b2             │               │ [Privilege escalation possible to other user than root]      │
│                    │                     │          │              │                         │               │ https://security-tracker.debian.org/tracker/TEMP-0841856-B1- │
│                    │                     │          │              │                         │               │ 8BAF                                                         │
```
## Related issues
- Close #3098 

## Related PRs
- [x] #3599 

Remove this section if you don't have related PRs.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/docs/target/container_image.md [hunk 1]
```diff
@@ -500,3 +500,10 @@ You can configure Docker daemon socket with `DOCKER_HOST` or `--docker-host`.
 ```shell
 $ trivy image --docker-host tcp://127.0.0.1:2375 YOUR_IMAGE
 ```
+
+### Configure Podman daemon socket to connect to.
+You can configure Podman daemon socket with `PODMAN_HOST` or `--podman-host`.
```
Comment (DmitriyLewen) [inline docs/docs/target/container_image.md:505]: It looks like `PODMAN_HOST` doesn't work

**Review by DmitriyLewen — COMMENTED** [review]
Hello @parvez0 
Thank you for your work!

Left a comment. Please take a look when you have time.

Can you also add a test for podman-host?
You can create a similar test:
https://github.com/aquasecurity/trivy/blob/c1d26ec3342ec776346c308b30e3a64d90effef1/pkg/fanal/image/daemon/image_test.go#L74-L116 

Regards, Dmitriy

File: docs/docs/target/container_image.md (hunk 1 above)
Comment (parvez0) [reply docs/docs/target/container_image.md:505]: Hey @DmitriyLewen thanks for reviewing the PR, I have added the missing test case as well as support for taking the PODMAN_HOST. Please do let me know if there are any fixes required.

File: pkg/fanal/image/daemon/image_test.go [hunk 2]
```diff
@@ -115,6 +115,52 @@ func Test_image_ConfigNameWithCustomDockerHost(t *testing.T) {
...
+
+		customDockerHost := filepath.Join(dir, "image-test-podman-socket.sock")
+		eo.UnixDomainSocket = customDockerHost
+		podmanSocket = customDockerHost
+	}
+
+	te := engine.NewDockerEngine(eo)
+	defer te.Close()
+
+	if runtime.GOOS == "windows" {
```
Comment (DmitriyLewen) [inline pkg/fanal/image/daemon/image_test.go:148]: Windows test is FAIL - https://github.com/aquasecurity/trivy/actions/runs/8145689974/job/22269334521?pr=6256#step:10:3751

Looks like we need to skip this test for windows (https://github.com/aquasecurity/trivy/blob/9361cdb7e28fd304d6fd2a1091feac64a6786672/pkg/fanal/image/daemon/podman_test.go#L45-L47).

Comment (parvez0) [reply pkg/fanal/image/daemon/image_test.go:148]: Hey, @DmitriyLewen, I need a clarification, what should happen if the test were run on Windows should the test pass by default?

**Review by DmitriyLewen — APPROVED** [review]
@parvez0 Thanks for your work!

@knqyf263 i approved this PR. Take a look, when you have time, please.

File: pkg/fanal/image/daemon/podman.go [hunk 3]
```diff
@@ -25,10 +25,15 @@ type podmanClient struct {
...
 }
 
-func newPodmanClient() (podmanClient, error) {
+func newPodmanClient(host string) (podmanClient, error) {
 	// Get Podman socket location
 	sockDir := os.Getenv("XDG_RUNTIME_DIR")
 	socket := filepath.Join(sockDir, "podman", "podman.sock")
+	if host != "" {
+		socket = host
+	} else if s := os.Getenv("PODMAN_HOST"); s != "" {
```
Comment (knqyf263) [inline pkg/fanal/image/daemon/podman.go:34]: `DOCKER_HOST` is natively supported by Docker.
https://docs.docker.com/engine/reference/commandline/cli/

But I didn't find PODMAN_HOST support in Podman. If Podman doesn't support it, there is no reason we do it.

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
LGTM! Thanks.

**Comment by knqyf263** [comment]
Sorry, I accidentally closed it... never mind.

**Pull request merged**